keps/sig-storage/2857-runtime-assisted-pv-mounts/README.md

```diff
@@ -0,0 +1,1767 @@
+- The mutating webhook analyzes the pod spec against the capabilities of the CSI
+plugin (based on the annotations from the CSIDriver) and RuntimeClass (based on
+the annotations in the RuntimeClass) and determines the CSI plugin (prior to
+upgrade) is not compatible. Based on this, the mutating webhook leaves the
+`FSDeferralStrategy` in the `PersistentVolumeClaimVolumeSource` of the pod spec
+empty.
+- The pod is scheduled to an appropriate node with the specified runtime
+installed while the PVC gets bound to a PV either prior to the scheduling or
+right after scheduling depending on volume binding mode of the storage class.
+- Kubelet sees `FSDeferralStrategy` is not specified for the pod and does not
+specify `fs_deferral_strategy` for any CSI APIs.
+- All file system operations associated with the CSI volume are managed by the
+CSI plugin and no deferral of any operations to the container runtime takes
+place even after the CSI plugin gets upgraded to support deferral of file system
+operations using the `kata` strategy.
+
+#### Story 12: Fallback to file system mount and management by CSI plugin due to CSI version downgrade after pod initialization resulting in failures
+
+All assumptions for the cluster mentioned earlier are true except the CSI plugin
+`plugin1.example.com` gets downgraded from a version that did support deferral
+of file system operations to a version that does not support deferral. The node
+did not get drained as part of the CSI node plugin downgrade.
+
+Prior to the downgrade of the CSI plugin, a pod is created in the cluster, with
+`runtimeClassName` set to `microvm` and PVCs with StorageClass set to
+`fast-storage`.
+
+The following sequence of steps takes place around mounting the file system
+on persistent volumes (bound to the PVCs specified by the pod) and applying
+post-mount configuration (specified by the pod):
+- The mutating webhook analyzes the pod spec against the capabilities of the CSI
+plugin (based on the annotations from the CSIDriver prior to downgrade) and
+RuntimeClass (based on the annotations in the RuntimeClass) and determines all
+of them are compatible. Based on this, the mutating webhook sets
+`FSDeferralStrategy` in the `PersistentVolumeClaimVolumeSource` of the pod spec
+to `kata`.
+- The PVC of the pod gets bound to a PV and the pod is scheduled to an
+appropriate node with the specified runtime installed.
+- The Kubelet invokes CSI `NodePublishVolume` passing the following based on the
+pod spec: `fs_deferral_strategy` set to `kata`, `volume_mount_group` set to
+`100`, `volume_mount_group_change_policy` set to `OnRootMismatch`.
+- The CSI plugin processes CSI NodePublishVolume by coordinating deferral of
+file system mount and application of post mount configuration steps with Kata.
+- The pod and all it's containers initialize into a running state within a micro
+VM with deferred file system mounts.
+
+While the pod runs, the CSI node plugin gets downgraded to a version of the CSI
```

Comment:
Please describe also the case when a Pod gets  `fs_deferral_strategy` set to `kata`, but is not running yet. CSI driver gets downgraded either before or when kubelet is mounting the volumes.

If I understood the proposal correctly, such Pod should fail to start (with a clear error message, right?) and needs to be deleted manually (?) and everything will be cleaned up.

Reply:
Added and yes.